File: Top_Interview_150/Sliding_Window/cpp/209.cpp
#include <vector>

using namespace std;

class Solution
{

    bool check_answer(vector<int> &nums, int n, int target, int k)
    {
        int s = 0;
        for (int i = 0; i < k; ++i)
            s += nums[i];

        if (s >= target)
            return true;

        for (int i = k; i < n; ++i)
        {
            s -= nums[i - k];
            s += nums[i];

            if (s >= target)
                return true;
        }

        return false;
    }

public:
    int minSubArrayLen(int target, vector<int> &nums)
    {
        int n = nums.size();

        int left = 0;
        int right = n;
        while (left <= right)
        {
            int middle = (left + right) / 2;
            if (this->check_answer(nums, n, target, middle))
                right = middle - 1;
            else
                left = middle + 1;
        }

        return left <= n ? left : 0;
    }
};

File: Top_Interview_150/Sliding_Window/python/209.py
from typing import List


class Solution:

    def check_answer(self, nums, target, k):
        s = 0
        for i in range(k):
            s += nums[i]

        if s >= target:
            return True

        for i in range(k, self.n):
            s -= nums[i - k]
            s += nums[i]

            if s >= target:
                return True

        return False

    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        self.n = len(nums)

        left = 0
        right = self.n
        while left <= right:
            middle = (left + right) // 2
            if self.check_answer(nums, target, middle):
                right = middle - 1
            else:
                left = middle + 1

        return left if left <= self.n else 0
